[Ticket #869bhrjg6] enhance link usage info with balance cache

baolongt opened the pull request:
Ticket https://app.clickup.com/t/869bhrjg6
- Backend
  - Extend AssetInfo model
    - add new amount_avalaible: Nat
  - operation update
    - create action: amount_avalaible = 0
    - when process action success, update for action
         - CreateLink: amount_avalaible += (amount_per_use + ledger_fee) * max_use 
         - Receive: amount_avalaible -= amount_per_use + ledger_fee
         - Send: amount_avalaible += amount_per_use
         - Withdraw: amount_avalaible = 0
- Frontend
  - LinkDto.assetInfo includes amount_avalaible field
  - Balance in usage info render from amount_avalaible

baolongt committed:
Merge remote-tracking branch 'origin/dev' into feature/show-link-balance

baolongt committed:
WIP

baolongt committed:
Merge remote-tracking branch 'origin/dev' into feature/show-link-balance

baolongt committed:
Merge remote-tracking branch 'origin/dev' into feature/show-link-balance

baolongt committed:
test: assert amount_available is correctly updated after withdraw and receive actions

baolongt committed:
feat: update version to 0.0.2 and refactor usageInfoSection to compute assetsWithTokenInfo from assetInfo

baolongt committed:
fmt

baolongt committed:
fmt

0xbarchitect commented on src/cashier_backend/src/apps/link_v2/links/shared/send_link/states/active.rs:
```diff
@@ -87,6 +90,23 @@ impl<M: TransactionManager + 'static> ActiveState<M> {
                 link.state = LinkState::InactiveEnded;
             }
+
+            // Decrement amount_available for each asset: amount_per_use + fee
```
also need to handle failure case

baolongt replied:
resolved

0xbarchitect commented on src/cashier_backend/src/apps/link_v2/links/shared/send_link/states/inactive.rs:
```diff
@@ -97,6 +97,11 @@ impl<M: TransactionManager + 'static> InactiveState<M> {
 
         if process_action_result.is_success {
```
also need to handle failure case

baolongt replied:
resolved

baolongt committed:
feat: Enhance multi-token link handling for partial success scenarios

- Updated ActiveState to decrement `amount_available` based on actual transferred amounts, handling partial successes.
- Modified CreatedState to set `amount_available` correctly when activating links, considering actual transfers.
- Adjusted InactiveState to manage `amount_available` during withdrawal actions, ensuring accurate balance updates.
- Introduced `actual_transferred_amounts` in ProcessActionResult to track successful transfers per asset.
- Added integration tests for partial success scenarios in multi-token basket links, verifying correct behavior during activation and receiving actions.
- Implemented utility functions for draining tokens from link subaccounts to simulate insufficient balance scenarios.
- Enhanced transaction manager to calculate successful transfer amounts from transactions, improving accuracy in multi-asset operations.

baolongt committed:
Merge remote-tracking branch 'origin/dev' into feature/show-link-balance

baolongt committed:
fix lint issue

baolongt commented on src/cashier_backend/src/apps/link_v2/links/shared/receive_link/states/active.rs:
```diff
@@ -73,12 +73,21 @@ impl<M: TransactionManager + 'static> ActiveState<M> {
             .process_action(action, intents, intent_txs_map)
             .await?;
 
+        if process_action_result.is_success {
```
resolved

baolongt commented:
Newest update 
- Add actual_transferred_amounts for process_action result. Any success transaction will added to this array
- Handle amount_available update when
  - Receive link with action Send, Withdraw
  - Send link with action CreateLink, Receive, Withdraw
- Add integration test for partial success transfer
  - CreateLink partial send
  - Withdraw + Receive partial reiceve

baolongt commented on src/cashier_backend/src/apps/link_v2/links/shared/receive_link/states/active.rs:
resolved

baolongt commented on src/cashier_backend/src/apps/link_v2/links/shared/send_link/states/created.rs:
```diff
@@ -97,6 +101,23 @@ impl<M: TransactionManager + 'static> CreatedState<M> {
         if process_action_result.is_success {
             link.state = LinkState::Active;
+
+            // Calculate and set amount_available for each asset
```
resolved

baolongt committed:
feat: Refactor transfer amount calculations and update link balance logic

0xbarchitect commented on src/cashier_backend/src/apps/link_v2/links/shared/receive_link/states/active.rs:
```diff
@@ -73,12 +74,28 @@ impl<M: TransactionManager + 'static> ActiveState<M> {
 
+        // Update amount_available based on actual transferred amounts (handles partial success)
+        // Send action: wallet -> link, so use wallet_to_link amounts
+        for asset_info in link.asset_info.iter_mut() {
```
this logic is duplicated across many states, such as this state, CREATED state of send-link.
implement it as an utils function to avoid code duplication.

0xbarchitect commented on src/cashier_backend/src/apps/link_v2/links/shared/receive_link/states/inactive.rs:
```diff
@@ -95,6 +98,31 @@ impl<M: TransactionManager + 'static> InactiveState<M> {
+        // Note: Withdraw action creates transfer for (balance - fee), so we need to add fee back
+        // Withdraw action: link -> wallet, so use link_to_wallet amounts
+        let fee_map = get_batch_tokens_fee_for_link(&link).await?;
+        for asset_info in link.asset_info.iter_mut() {
```
this logic is duplicated across many states, such as this state, ACTIVE and INACTIVE states of send-link.
implement it as an utils function to avoid code duplication.

0xbarchitect commented on src/cashier_backend/src/apps/link_v2/links/shared/receive_link/states/inactive.rs:
```diff
@@ -95,6 +98,31 @@ impl<M: TransactionManager + 'static> InactiveState<M> {
+        // Decrement amount_available based on actual withdrawn amounts (handles partial success)
+        // Note: Withdraw action creates transfer for (balance - fee), so we need to add fee back
+        // Withdraw action: link -> wallet, so use link_to_wallet amounts
+        let fee_map = get_batch_tokens_fee_for_link(&link).await?;
```
this lookup introduces latency, making the processing action slower, which we are trying to avoid.
is there better solution?

0xbarchitect commented on src/cashier_backend/src/apps/link_v2/links/shared/send_link/states/active.rs:
```diff
@@ -82,6 +85,31 @@ impl<M: TransactionManager + 'static> ActiveState<M> {
 
+        // Decrement amount_available based on actual transferred amounts (handles partial success)
+        // Use action: link -> wallet, so use link_to_wallet amounts
+        let fee_map = get_batch_tokens_fee_for_link(&link).await?;
```
same comments as above

0xbarchitect commented on src/cashier_backend/src/apps/link_v2/links/shared/send_link/states/created.rs:
```diff
@@ -94,7 +95,21 @@ impl<M: TransactionManager + 'static> CreatedState<M> {
-        // if process action succeeds, activate the link
+        // Update amount_available based on actual transferred amounts (handles partial success)
+        // Create action: wallet -> link, so use wallet_to_link amounts
+        for asset_info in link.asset_info.iter_mut() {
```
same comment as above

0xbarchitect commented on src/cashier_backend/src/apps/link_v2/links/shared/send_link/states/inactive.rs:
```diff
@@ -95,6 +98,31 @@ impl<M: TransactionManager + 'static> InactiveState<M> {
+        // Decrement amount_available based on actual withdrawn amounts (handles partial success)
+        // Note: Withdraw action creates transfer for (balance - fee), so we need to add fee back
+        // Withdraw action: link -> wallet, so use link_to_wallet amounts
+        let fee_map = get_batch_tokens_fee_for_link(&link).await?;
```
same comment as above

0xbarchitect commented on src/cashier_backend_types/src/link_v2/action_result.rs:
```diff
@@ -54,6 +111,9 @@ pub struct ProcessActionResult {
     pub errors: Vec<String>,
+    /// Transfer amounts grouped by direction (wallet_to_link vs link_to_wallet)
+    /// Calculated from successful transactions based on intent task type
+    pub transfer_amounts: TransferAmountsResult,
```
the `transfer_amounts` is derived values from processed intents and transactions.
don't introduce it as a new field in the ProcessActionResult, instead calculating it from result data, using an utils pure  function.

0xbarchitect commented on src/cashier_frontend_new/src/modules/creationLink/types/createLinkData.ts:
because the PR is big, we can split it into separate PRs for BE and FE to make it easier to review.
This PR should be focused on BE only, and ignore the FE for now.

0xbarchitect commented on src/cashier_frontend_new/src/modules/detailLink/components/usageInfoSection.svelte:
same comment as above

0xbarchitect commented on src/cashier_frontend_new/src/modules/detailLink/pages/detail.svelte:
same comment as above

0xbarchitect commented on src/cashier_frontend_new/src/modules/detailLink/utils/transaction-lock-status.ts:
same comment as above

0xbarchitect commented on src/cashier_frontend_new/src/modules/links/types/link/asset.ts:
same comment as above

0xbarchitect commented on src/transaction_manager/src/ic_transaction_manager.rs:
```diff
@@ -210,13 +211,20 @@ impl<E: IcEnvironment> TransactionManager for IcTransactionManager<E> {
             )?;
 
+            // Calculate transfer amounts grouped by direction (wallet_to_link vs link_to_wallet)
+            let transfer_amounts = calculate_successful_transfer_amounts(
```
the `transfer_amounts` is a derived value from processed intents and transactions, don't introduce it as a new field in ProcessActionResult and implement it in transaction_manager, instead calculating it in link state handler logic.

0xbarchitect commented on src/transaction_manager/src/ic_transaction_manager.rs:
```diff
@@ -210,13 +211,20 @@ impl<E: IcEnvironment> TransactionManager for IcTransactionManager<E> {
                 icrc112_requests: Some(icrc112_requests),
                 is_success,
                 errors,
+                transfer_amounts,
```
same comment as above

0xbarchitect commented on src/transaction_manager/src/utils/calculator.rs:
please move this calculator file to `cashier_backend/apps/link_v2/utils/...`
it is wrongly placed here in transaction_manager package, because we want to make transaction_manager link-agnostic and supposed to handle intents and transactions only.
all the links related logic should be implemented in `cashier_backend` package

0xbarchitect commented on src/cashier_backend/src/apps/link_v2/links/shared/receive_link/actions/send.rs:
```diff
@@ -52,12 +52,8 @@ impl SendAction {
-        let token_fee_map = get_batch_tokens_fee_for_link(link).await?;
-
-        let link_token_balance_map = calculate_link_balance_map(
+        let link_token_balance_map = calculate_link_balance_map_without_fee(
```
for Payment link, the Payer doesn't need to cover the ledger fee?
